authentication/oidc.go

```diff
@@ -45,71 +47,112 @@ type OIDCConfig struct {
+type OIDCHandlers struct {
+	mtx         sync.RWMutex
+	handlers    map[string]http.Handler
+	middlewares map[string]Middleware
 
+	logger          log.Logger
+	retryCountIncFn func()
+}
+
+// NewOIDCHandlers instantiates OIDC handlers.
+func NewOIDCHandlers(l log.Logger, reg prometheus.Registerer) *OIDCHandlers {
 	tenantErrors := prometheus.NewGauge(prometheus.GaugeOpts{
 		Namespace: "observatorium",
 		Subsystem: "api",
 		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Help:      "The number of attempts for which tenant OIDC provider instantiation failed.",
 	})
 
 	reg.MustRegister(tenantErrors)
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
+	return &OIDCHandlers{
+		handlers:        make(map[string]http.Handler),
+		middlewares:     make(map[string]Middleware),
+		logger:          l,
+		retryCountIncFn: tenantErrors.Inc,
+	}
+}
 
-			tenantErrors.Inc()
+// AddOIDCForTenant adds a new handler and middleware for a new tenant, based on
+// the tenant OIDC configuration. If a connection with the OIDC provider
+// cannot be established, the method will retry to connect with an increasing
+// backoff.
+func (oh *OIDCHandlers) AddOIDCForTenant(prefix string, config TenantOIDCConfig) {
+	ctx := context.Background()
+	b := backoff.New(ctx, backoff.Config{
+		Min:        500 * time.Millisecond,
+		Max:        5 * time.Second,
+		MaxRetries: 0, // retry indefinitely
+	})
 
-			continue
-		}
+	go func() {
+		for b.Reset(); b.Ongoing(); {
+			p, err := NewProvider(ctx, oh.logger, getCookieForTenant(config.Tenant), "/"+config.Tenant, config.OIDCConfig)
+			if err != nil {
+				level.Warn(oh.logger).Log("msg", "failed to instantiate OIDC provider for tenant", "tenant", config.Tenant, "error", err)
+				oh.retryCountIncFn()
+				b.Wait()
+				continue
+			}
 
-		r := chi.NewRouter()
+			r := chi.NewRouter()
 
-		const (
-			loginRoute    = "/login"
-			callbackRoute = "/callback"
-		)
+			const (
+				loginRoute    = "/login"
+				callbackRoute = "/callback"
+			)
 
-		r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
-		r.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
+			r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
+			r.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
 
-		handlers[c.Tenant] = r
-		middlewares[c.Tenant] = p.Middleware()
-	}
+			oh.mtx.Lock()
+			oh.handlers[config.Tenant] = r
+			oh.middlewares[config.Tenant] = p.Middleware()
+			oh.mtx.Unlock()
+
+			level.Debug(oh.logger).Log("msg", "OIDC provider instantiated for tenant", "tenant", config.Tenant)
```

Comment:
nit: I would make this `info` level - it should only happen at container start-up once

Reply:
With retries, it will log for every retry (every 5 sec in the best case), considering this, `debug` looks better to me.

Reply:
Yes, that makes sense actually, it is not too spammy and useful.